@isTest
public class CaseTriggerHandlerTest {
    @isTest private static void preventCreateTest(){
        Account testAccount = new Account(Name = 'Test');
        insert testAccount;
        Case caseTestInsert = new Case(AccountId = testAccount.Id, Origin = 'Email', Status='New', Subject='Test2');
        insert caseTestInsert;
        Case testCase = new Case(AccountId = testAccount.Id, Origin = 'Email', Status='New', Subject='Test2');
        Database.SaveResult sr = Database.insert(testCase, false);
        System.assert(!sr.isSuccess());
        System.assert(sr.getErrors().size() > 0);
        
        for(Database.Error err: sr.getErrors()){
            System.assertEquals(System.StatusCode.FIELD_CUSTOM_VALIDATION_EXCEPTION, err.statuscode);
            System.assert(err.getMessage().contains('You can`t create a case with the same subject'));
        } 
    }
    @isTest private static void preventInsertArray(){
        Account testAccount = new Account(Name = 'TestArray');
        insert testAccount;
        List<Case> caseList = new List<Case>();
        for(Integer i = 0 ; i < 10 ; i++){
            Case newTestCases = new Case(AccountId = testAccount.Id, Origin = 'Email', Status='New', Subject='TestArray');
            caseList.add(newTestCases);
        }
        Database.insert(caseList, false);
        List <Case> selectList = [SELECT Id FROM Case];
        System.assertEquals(1, selectList.size());   
    }
    @isTest private static void preventUpdateReccord(){
        Account testAccount = new Account (Name='TestUpdate');
        insert testAccount;
        Case firstTestCase = new Case (AccountId = testAccount.Id, Origin = 'Email', Status='New', Subject='TestUpdateCase');
        insert firstTestCase;
        Case secondTestCase = new Case (AccountId = testAccount.Id, Origin = 'Email', Status='New', Subject='TestUpdateCase2');
        insert secondTestCase;
        secondTestCase.Subject = 'TestUpdateCase';
        Database.SaveResult sr = Database.update(secondTestCase, false);
        System.assert(!sr.isSuccess());
        System.assert(sr.getErrors().size() > 0);
        
        for(Database.Error err: sr.getErrors()){
            System.assertEquals(System.StatusCode.FIELD_CUSTOM_VALIDATION_EXCEPTION, err.statuscode);
            System.assert(err.getMessage().contains('You can`t create a case with the same subject'));
        } 
    }
}
